src/PostgREST/App.hs

```diff
@@ -1,100 +1,130 @@
+          Right (qs, cqs) -> do
+            let qt = qualify table
+                count = if hasPrefer "count=none"
                       then countNone
-                      else countRows qt
-            query = B.Stmt "select " V.empty True <>
-                parentheticT (
-                  whereT qt qq count
-                ) <> commaq <> (
-                bodyForAccept contentType qt
-                . limitT range
-                . orderT (orderParse qq)
-                . whereT qt qq
-                $ select qt qq
-              )
-        row <- H.maybeEx query
-        let (tableTotal, queryTotal, body) =
-              fromMaybe (Just 0, 0, Just "" :: Maybe Text) row
-            to = from+queryTotal-1
-            contentRange = contentRangeH from to tableTotal
-            status = rangeStatus from to tableTotal
-            canonical = urlEncodeVars
-              . sortBy (comparing fst)
-              . map (join (***) cs)
-              . parseSimpleQuery
-              $ rawQueryString req
-        return $ responseLBS status
-          [contentTypeH, contentRange,
-            ("Content-Location",
-              "/" <> cs table <>
-                if Prelude.null canonical then "" else "?" <> cs canonical
-            )
-          ] (cs $ fromMaybe "[]" body)
+                      else cqs
+
+                q = B.Stmt "select " V.empty True <>
+                    parentheticT count
+                    <> commaq <> (
+                    bodyForAccept contentType qt -- TODO! when in csv mode, the first row (columns) is not correct when requesting sub tables
+                    . limitT range
+                    $ qs
+                  )
+            -- return $ responseLBS status200 [contentTypeH] (cs $ show $ B.stmtTemplate q)
+            row <- H.maybeEx q
+            let (tableTotal, queryTotal, body) = fromMaybe (Just (0::Int), 0::Int, Just "" :: Maybe Text) row
+                to = from+queryTotal-1
+                contentRange = contentRangeH from to tableTotal
+                status = rangeStatus from to tableTotal
+                canonical = urlEncodeVars
+                  . sortBy (comparing fst)
+                  . map (join (***) cs)
+                  . parseSimpleQuery
+                  $ rawQueryString req
+
+
+            return $ responseLBS status
+              [contentTypeH, contentRange,
+                ("Content-Location",
+                  "/" <> cs table <>
+                    if Prelude.null canonical then "" else "?" <> cs canonical
+                )
+              ] (cs $ fromMaybe "[]" body)
+
+        where
+            from = fromMaybe 0 $ rangeOffset <$> range
+            apiRequest = first formatParserError (parseGetRequest req)
+                     >>= addRelations schema allRelations Nothing
+                     >>= addJoinConditions schema allColumns
+                     where formatParserError = pack.show
```

Comment:
Use the general `cs` function rather than `pack` since it can figure out what to do.